doc/manuals/tools/pipe-config.rst
===========
pipe-config
===========

The pipe-config applet generates a new pipeline file from a set of
inputs. All the information from the config files and settings
is merged into the specified pipe file and written as output.
This can be used to flatten a complicated pipe file which may include
other pipe and config files into a single dasily distributed file.


kwiver pipe-config       [options] pipe-file
--------------------------------------------

**Options are:**

  ``-h, --help``

    Display applet usage information.

**output options:**

  ``-o, --output arg``

    Name of output file or '-' for stdout. (default: -)

**pipe options:**

  ``-c, --config <file>``

    Specifies a file containing supplemental configuration entries.
    Can occur multiple times on the command line.

    The supplemental configuration files can contain additional
    configuration entires for processes or global config
    blocks. Configuration items for processes start with the process
    name. For example The ``input_path`` config element for the process
    ``reader`` would be specified as follows:

    ``reader:input_path = input/file/path``

    These configuration values are applied to the pipeline configuration
    after the pipeline file has been processes and can overwrite values
    previously defined in the pipe file.

  ``-s, --settings VAR=VALUE``

    Specifies additional configuration settings to be applied to the pipeline configuration.

    The settings option can be used to specify a single configuration
    value. This command line option can appear multiple times on the
    command line. Using the same example above, that configuration item can be
    specified as follows:

    ``--settings reader:input_path=input/file/path``

    Note that there can be no embedded spaces in the argument.

  ``-I, --include <path>``

    Specifies a directory to be added to the configuration include path.
    This option can occur multiple times on the command line

  pipe-file  - name of pipeline file.


doc/manuals/vital/config_usage.rst
Configuration Usage
===================

Introduction
------------

The vital ``config_block`` supports general configuration tasks where
a general purpose key/value pair is needed. The configuration block is
used to specify, communicate and set configurable items in many
different situations. The two major users of the configuration support
are algorithms and processes. In addition, there are a few other
places that they are used also.

Configurations are ususally established in an external file which are
read and converted to an internal ``config_block`` object. This is the
typical way to control the behaviour of the software. Configuration
blocks can also be created pragmatically such as when specifying an
expected set of configurable items.

When algorithms are used within processes, the


From File to config_block
-------------------------

Using config_block_io to directly convert config file into block. This
can be used by a main program that manages configs and algorithms
directly. The ``read_congfig_file()`` uses a complex set of rules to
locate config files based on host system and application name.


Configuration Features
----------------------

config features. what they do and why you would want to use them
- relativepath

- macro providers and how they can be used to make portable and reusable config files

- config sub-blocks and configuration context


Establishing Expected Config
----------------------------

Typically the expected config is formulated by creating a config block
with all the expected keys, default values, and entry
description. This is done for both algorithms and processes.

Don't be shy with the entry description. This description serves as
the design specification for the entry. The expected format is a short
description followed by a longer detailed description separated by two
new-lines, as shown below.

.. code-block:: c++

  config->set_value( "config_name", <default_value>,
                     "Short description.\n\n"
                     "Longer description which contains all information needed "
                     "to correctly specify this parameter including any range "
                     "limitations etc." );

The long description does not need any new-line characters for
formatting unless specific formatting is desired. The text is wrapped
into a text block by all tools that display it.

This expected configuration serves as documentation for the algorithm
or process configuration items when it is displayed by the
plugin_explorer and other tools. It is also used to validate the
configuration supplied at run time to make sure all expected items are
present.


Usage by Algorithms
'''''''''''''''''''

Algorithms specify their expected set of configurable items in their
``get_configuration()`` method using the config_block ``set_value()`` method,
described above.

The run time configuration is passed to an algorithm through the
``set_configuration()`` method. This method typically extracts the
expected configuration values and saves them locally for the algorithm
to use. When a configuration is read from the file, there is no
guarantee that all expected configuration items are present and
attempting to get a value that is not present generates an exception.

The recommended way to avoid this problem is to use the expected
configuration, as created by the ``get_configuration()`` method to
supply any missing entries. The following code snippet shows how this
is done.

.. code-block:: c++

    // Set this algorithm's properties via a config block
    void
    <algorithm>
    ::set_configuration(vital::config_block_sptr in_config)
    {
      // Starting with our generated vital::config_block to ensure that assumed values are present
      // An alternative is to check for key presence before performing a get_value() call.
      vital::config_block_sptr config = this->get_configuration();

      // Merge in supplied config to cause these values to overwrite the defaults.
      config->merge_config(in_config);

      // Get individual config entry values
      this->write_float_features = config->get_value<bool>("write_float_features",
                                                           this->write_float_features);
    }



Instantiating Algorithms
''''''''''''''''''''''''

Algorithms can be used directly in application code or they can be
wrapped by a sprokit process. In either case the actual implementation
of the abstract algorithm interface is specified through a config block.

Lets first look at the code that will instantiate the configured
algorithm and then look at the contents of the configuration file.

The following code snippet instantiates a ``draw_detected_object_set``
algorithm.

.. code-block:: c++

  // this pointer will be used to reference the algorithm after it is created.
  vital::algo::draw_detected_object_set_sptr m_algo;

  // Get algorithm configuration
  auto algo_config = get_config(); // or an equivalent call

  // Check config so it will give run-time diagnostic of config problems
  if ( ! vital::algo::draw_detected_object_set::check_nested_algo_configuration( "draw_algo", algo_config ) )
  {
    LOG_ERROR( logger, "Configuration check failed." );
  }

  vital::algo::draw_detected_object_set::set_nested_algo_configuration( "draw_algo", algo_config, m_algo );
  if ( ! d->m_algo )
  {
    LOG_ERROR( logger, "Unable to create algorithm." );
  }

After the configuration is extracted, it is passed to the
``check_nested_algo_configuration()`` method to determine if the
configuration has the basic ``type`` entry and the requested type is
available. If the ``type`` entry is missing or the specified
implementation is not available, a detailed log message is generated
with the available implementations.

If the configuration is acceptable, the
``set_nested_algo_configuration()`` call will actually instantiate and
configure the selected algorithm implementation.

The name that is supplied to these calls, "draw_algo" in this case, is
used access the configuration block for this algorithm.

The following configuration file snippet can be used to configure
the above algorithm.::

  block draw_algo
    type = ocv    # select the ocv instance of this algorithm

    block ocv     # configure the 'ocv' instance
      alpha_blend_prob   = true
      default_line_thickness   = 1.25
      draw_text   = false
    endblock # for ocv
  endblock  # for draw_algo

The outer block labeled "draw_algo" specifies the configuration to be
used for the above code snippet. The config entry "type" specifies
which implementation of the algorithm to instantiate. The following
block labeled "ocv" is used to configure the algorithm after it is
instantiated. The block labeled "ocv" is used for algorithm type
"ocv". If the algorithm type was "foo", then the block "foo" would be
used to configure the algorithm.


Usage by Processes
''''''''''''''''''

The configuration for sprokit processes is presented slightly
differently than for algorithms, but underneath, they both use the
same structure.

Configuration items for a process are defined using
``create_config_trait()`` macro as shown below.

.. code-block:: c++

  //                    name,      type,  default,        description
  create_config_trait( threshold, float, "-1", "min threshold for output (float).\n\n"
                       "Detections with confidence values below this threshold are not drawn." );

When the process is constructed all configuration parameters must be
declared using the ``declare_config_using_trait()`` call, as shown below.::

  declare_config_using_trait( threshold );

All configuration items declared in this way are available for display
using the plugin_explorer tool.

Configuration values are extracted from the process configuration in
the ``_configure()`` method of the process as shown below.::

  float local_threshold = config_value_using_trait( threshold );

Processes can instantiate and configure algorithms using the approach
described above.

Configuration for a process comes from a section of the pipe file. The
following section of a pipe file shows configuration for a process
which supplies the threshold configuration item.::

  # ================================
  process draw_boxes :: draw_detected_object_boxes
    threshold = 3.14


Verifying a Configuration
'''''''''''''''''''''''''

When a configuration file (or configuration section of a pipe file) is
read in, there is no checking of the configuration key names. There is
no way of knowing which configuration items are valid or expected and
which ones are not. If a name is misspelled, which sometimes happens,
it will be misspelled in the configuration block. This can lead to
hours of frustration diagnosing a problem.

A configuration can be checked against a baseline using the
config_difference class. This class provides methods to determine the
differences between a reference configuration and one created from an
input file. The difference between these two configurations is
presented in two different ways. It provides a list of keys that are
baseline config and not in the supplied config. These are the config
items that were expected but not supplied. It also provides a list of
keys that are in the supplied config but not in the expected
config. These items are supplied but not expected.

The following code snippet shows how to report the difference between
two config blocks.

.. code-block:: c++

  //                                    ref-config                received-config
  kwiver::vital::config_difference cd( this->get_configuration(), config );
  const auto key_list = cd.extra_keys();
  if ( ! key_list.empty() )
  {
    // This may be considered an error in some cases
    LOG_WARN( logger(), "Additional parameters found in config block that are not required or desired: "
              << kwiver::vital::join( key_list, ", " ) );
  }

  key_list = cd.unspecified_keys();
  if ( ! key_list.empty() )
  {
    LOG_WARN( logger(), "Parameters that were not supplied in the config, using default values: "
              << kwiver::vital::join( key_list, ", " ) );
  }


Not all applications need to check both cases. There may be good
reasons for not specifying all expected configuration items when the
default values are as expected. In some cases, unexpected items that
are supplied by the configuration may be indications of misspelled
entries.


Config Management Techniques
----------------------------

The configuration file reader provides several alternatives for
managing the complexity of a large configuration. The block / endblock
construct can be used to shorted config lines and modularize the
configuration. The include directove can be used to share or reuse
portions of a config.

Starting with the example config section that selects an algorithm and
configures it::

    algorithm_instance_name:type = type_name
    algorithm_instance_name:type_name:algo_param = value
    algorithm_instance_name:type_name:threshold = 234

The block construct can be used to simplify the configuration and
make it easier to navigate.::

  block algorithm_instance_name
    type = type_name
    block  type_name
      algo_param = value
      threshold = 234
    endblock
  endblock

In cases where the configuration block is extensive or used in
multiple applications, that part of the configuration can exist as a
stand-alone file and be included where it is needed.::

    block  algorithm_instance_name
      include type_name.conf
    endblock

where ``type_name.conf`` contains::

    type = type_name
    block   type_name
      algo_param = value
      threshold = 234
    endblock

Environment variables and config macros can be combined to provide a
level of adaptability to config files. Using the environment macro in
an include directive can provide run time agility without requiring
the file to be edited. The following is an example of selecting a
different include file based on mode.::

  include $ENV{MODE}/config.file.conf


RELEASE_NOTES.md
v0.10.6 - 3/12/2020
===================


-Fix scoring examples which became broken in last


-Add tracker training example for windows binaries


-Fix issue in running test-time multi-GPU from an image input list



v0.10.5 - 2/26/2020
===================


-Add improved CFRNN training proceedure using netharn



v0.10.4 - 1/31/2020
===================


-Fix issue in centos binaries backing X11 library



v0.10.3 - 1/13/2020
===================


-Refine cascade faster rcnn training proceedure first pass


-Add tracker-assisted annotation for video


-Fix split track option in default annotator


-New seal multi-view GUI release



v0.10.2 - 12/1/2019
===================


-Fix CentOS issues with unable to run pipelines (Ubuntu and Windows fine)



v0.10.2 - 11/25/2019
====================


-Updated KWIVER in VIAME causing all add-ons to break and need to be regenerated


-Added in support for object tracking solely based on IOU and image stabilization


-Fixed issue in installers with measurement code after upgrading numpy versions


-Add simple mosaic generation script and related pipelines


-Add multiple seal tracking example pipelines


-Fix issues in prior v0.10.* releases on systems with no CUDA 10s installed



v0.10.1 - 11/16/2019
====================


-Remove ball-tree indexing in favor of full LSH hash searches due to issues


-Begin packaging python in binaries for linux in addition to windows



v0.10.0 - 11/15/2019
====================


-Baseline speed optimizations and bug fixes for IQR, significant database speedups


-Begin packaging python in binaries for windows, will do this for linux soon



v0.9.18 - 11/01/2019
===================


-Fixed windows issues with running and training detectors introduced in last


v0.9.17 - 10/18/2019
====================


-Quick fixes to make IQR GUI not crash on images of different size and large images


-Add SVM training tool example, for training SVMs from large amounts of annotations


-Misc tuning and bug fixes



v0.9.16 - 7/24/2019
===================


-Lots of bug fixes (windows project files)


-Learning rate tuning for latest detection models


-CUDA 10.0 support



v0.9.13 - 5/30/2019
===================


-Add missing .so files into CentOS and Ubuntu binaries to not require external packages



v0.9.12 - 5/21/2019
===================


-Support local detection running for multiple variants of motion detectors


-All detector training routines now also generate KWIVER pipelines


-Remove index_existing and merged changes into index_defaults to reduce confusion


-Additional detection models



v0.9.11 - 5/16/2019
===================


-New default and generic object detection models


-Improved IQR detector performance


-The default box detector training routine is now cascade rnn instead of yolo



v0.9.10.0 - 4/22/2019
=====================


-New detection techniques and default models: cascade faster rcnn, cfrnn with motion, YOLO-WTF


-Additional support for frame-level classification in the system


-Image registration example added, arctic seal examples



v0.9.9.7 - 3/19/2019
====================


-Fixed a few issues in last relating to detector optimizations which broke things in last


-More intuitive final model saveouts


v0.9.9.6 - 2/25/2019
====================


-Windows 7 runtime training tool fixes


-Better automatic video detection, don't process hidden files, dirs recursively or non-videos


-Show correct time offsets in query GUI from video start


v0.9.9.5 - 2/21/2019
====================


-Fix bug added to deep training tool in last version


-Fix training tool windows error reporting


v0.9.9.4 - 2/18/2019
====================


-Fix model saveout in windows binaries broken in last


-Updated default model training parameters


v0.9.9.3 - 2/13/2019
====================


-Minor python change to drop dependency in standard use case


v0.9.9.2 - 2/12/2019
====================


-Add train deep model from GUI dropdown


v0.9.9.1 - 2/6/2019
===================


-Fix issues in prior release dealing with default indexing option and finding local models


v0.9.9.0 - 2/4/2019
===================


-Add YOLOv3, YOLOv3 with Temporal Features detectors


-GUI bug fixes: issues with running embedded pipes and detections not always showing up


-Configuration file cleanup


-Remove multiple models from default installers, move to seperate .zips



v0.9.8.11 - 12/4/2018
=====================


-Improved timeline plotting for multi-video inputs


-Fixed issue in image list project files introduced in prior


-More intelligent default track classification probabilities which make use of keyframes



v0.9.8.10 - 11/26/2018
======================


-Add automatic dynamic range compute for 16-bit imagery in GUI


-Expose video rate for image list processing in default project files


-Added extra error checking on indexing scripts involving databases, improved error reporting



v0.9.8.9 - 11/12/2018
=====================


-Allow video projects to process directory of directory of images


-Additional error checking on index formation


-Improvements to launch_timeline_viewer script


-Added GUI display dynamic range parameters for 16-bit imagery



v0.9.8.8 - 10/29/2018
=====================


-Fix issue with measurement example after config refactor



v0.9.8.7 - 10/25/2018
=====================


-Additional bug fixes and parameter tuning relating to tile boundaries when chipping enabled in detectors



v0.9.8.6 - 10/24/2018
=====================


-Fixed a bug in debayer and color correction filter .pipe which was causing garbled results in the example


-Added RELEASE_NOTES to binaries


-Fixes to Windows CPU-only binaries


-Fixes for properly processing greyscale imagery


-Non-maximum suppression tuning


-If there is a corrupt frame in the middle of a sequence, keep on processing in both GUI and CLI



v0.9.8.5 - 10/18/2018
=====================



plugins/pytorch/mdnet/options.py

import os

from collections import OrderedDict

viame_dir = os.environ['VIAME_INSTALL']
models_dir = os.path.join( 'configs', 'pipelines', 'models' )
model_file = os.path.join( viame_dir, models_dir, 'mdnet_seed.pth' )

opts = OrderedDict()

opts['use_gpu'] = True
opts['model_seed'] = model_file
opts['savefig_dir'] =''
opts['display']=False
opts['visual_log']=True

opts['crop_size'] = 107
opts['padding'] = 1.2
opts['jitter'] = True
opts['result_path']='./result.npy'
opts['adaptive_align']=True
opts['batch_pos'] = 32
opts['batch_neg'] = 96
opts['batch_neg_cand'] = 1024
opts['batch_test'] = 256

opts['n_samples'] = 256
opts['trans_f'] = 0.6
opts['scale_f'] = 1.05
opts['trans_f_expand'] = 1.4

opts['n_bbreg'] = 1000
opts['overlap_bbreg'] = [0.6, 1]
opts['scale_bbreg'] = [1, 2]

opts['lr_init'] = 0.0001 # original = 0.0001
opts['maxiter_init'] = 50 # original = 30
opts['n_pos_init'] = 500
opts['n_neg_init'] = 5000
opts['overlap_pos_init'] = [0.7, 1]
opts['overlap_neg_init'] = [0, 0.5]

opts['lr_update'] = 0.0003 # original = 0.0002
opts['maxiter_update'] = 15 # original = 15
opts['n_pos_update'] = 50
opts['n_neg_update'] = 200
opts['overlap_pos_update'] = [0.7, 1]
opts['overlap_neg_update'] = [0, 0.3]

opts['success_thr'] = 0. # original = 0
opts['n_frames_short'] = 20
opts['n_frames_long'] = 100
opts['long_interval'] = 10

opts['w_decay'] = 0.0005 # original = 0.0005
opts['momentum'] = 0.9
opts['grad_clip'] = 10 # original = 10
opts['lr_mult'] = {'fc6':10}
opts['ft_layers'] = ['fc']
